### ISSUE TITLE
Confusing/nonsensical errors when no build configuration exists.

### ISSUE THREAD

**Issue by alexchandel** [build2/build2#118]
[When following the tutorial's example of 1 exe and 1 lib package in the same overarching project](https://build2.org/build2-toolchain/doc/build2-toolchain-intro.xhtml#guide-dev-multi),

```sh
bdep new -t empty hello
bdep new --package -l c++ -t lib libhello
bdep new --package -l c++ -t exe hello
```

uncommenting the import line in `hello/hello/buildfile`,
```
import libs += libhello%lib{hello}
```

including `<libhello/hello.hxx>` in `hello/hello/hello.cxx`, and building in-place yields the nonsensical error:
```sh
$ b
error: unable to import target libhello%lib{hello}
  info: use config.import.libhello configuration variable to specify its project out_root
  info: while applying rule cxx.link to update hello/hello/exe{hello}
  info: while applying rule alias to update hello/dir{hello/}
  info: while applying rule alias to update dir{hello/}
  info: while applying rule alias to update dir{./}
version.in libhello/libhello/version.hxx.in
info: failed to update dir{./}
```

However it creates build artifacts in `libhello/libhello`.  These are difficult to remove, because `b clean` (and even `b clean libhello/`) fails with the same error.  `bdep init -a` doesn't fix anything either.  `bdep init -C ../hello-out @out` was required.

Can a more intelligent error message be provided in such cases?  (And how can you actually do an in-place build?)

**Comment by boris-kolpackov** [comment]
> building in-place yields the nonsensical error:

You've created two packages and then invoked the build system to build them (via the glue `buildfile` found in the root of your project). Since from the build system's point of view nothing connects them, `hello` fails to import `libhello`.

> These are difficult to remove

This works for me (notice `:` after `clean`):

```
b clean: libhello/
```

Again, if you are not sure whether something is a bug, it could be better to first check on slack first.

**Comment by alexchandel** [comment]
But the out-of-tree build configuration, which builds successfully, has nothing more connecting them.  Both have buildfile glue (`import pkgs = */` `./: $pkgs` or `./: {*/ -build/}`).  And `b --dump load` appears to show both of the packages resolved.

The bug is unclear error messages and documentation.  The error messages & documentation don't explain what's happening (or any hints on how solve it) in a manner comprehensible to someone who isn't already a `build2` expert (defeating the purpose).

**Comment by boris-kolpackov** [comment]
> But the out-of-tree build configuration, which builds successfully, has nothing more connecting them.

The build configuration (which itself is a project) connects them. An equivalent (to the in-source build) invocation would be something like this:

```
b hello/@../hello-out/ libhello/@../libhello-out/
```

Try it, and it will fail in exactly the same way.

> The bug is unclear error messages and documentation. The error messages & documentation don't explain what's happening (or any hints on how solve it) in a manner comprehensible to someone who isn't already a `build2` expert (defeating the purpose).

I empathize with your frustration, the learning curve is steep, I am not denying it. And we've also tried hard to explain the underlying mental model to the best of our abilities. So unless you have specific suggestions (or, better yet, patches) on how to improve this, I am not sure what else we can do here right now. Maybe we just need to wait and gather more feedback from more users to see if there is a common pattern of misunderstanding/confusion that can we can clarify.

**Comment by kamrann** [comment]
Just wanted to jump in and say that I've had a lot of the same issues in understanding. I feel like most of my successes have been more from trial and error than from reaching a higher general understanding of the concepts. But obviously it takes time, hopefully things will click at some point.

Anyway I think one central thing that it may be worth focusing on with the docs is precisely what a 'project' means in the the context of b2. I know there is a reference somewhere in the docs to the word's multiple meanings in general, but I think it's critical to be clear on what it means when used in the documentation, and if it always means exactly the same thing. The comment above, 'The build configuration (which itself is a project)' sounds important, but I honestly don't know quite what it's saying.

More generally, the main roadblock for me has been hard to decipher diagnostics, as Alex mentions. Specifically, error messages that say what the problem is from the point of view of the build process, but leaving the user to guess as to the root cause. "error: unable to import target" and "error: no rule to update" being the two most common ones. I know this is a hard problem to solve, but I've been gathering some notes on particular cases where it feels like it should be possible to give more hints as to likely cause; I'll get them to you at some point.

I also had the idea yesterday that some kind of online b2 project sandbox, where you could manipulate project structure and build files and see some sort of live(-ish) feedback on buildfile validity/project state/build invocation results would be a huge help for learning. Obviously it's potentially not trivial, but it could start small. That's something I'd happily try to contribute to.

**Comment by boris-kolpackov** [comment]
> I feel like most of my successes have been more from trial and error than from reaching a higher general understanding of the concepts. But obviously it takes time, hopefully things will click at some point.

I think one of the reasons for this experience is that people are not taking it slow, so to speak. And the reason for this, IMO, is that they are trying to handle complex projects right away with their own idiosyncrasies dictated by existing/previous build systems and/or their preferences of "how things should be" based on experience using other build systems.

At the same time, `build2` does not take the "black box" approach where one has a bunch of magic incantations for each supported setup. Rather, it tries to define a conceptual model of build that can then be used to implement various setups. The advantage is that it can support a much wider variety of setups (plus there is no magic). But there is also a disadvantage: the documentation and diagnostics is necessarily fairly general since we can't assume to know exactly what you would want to achieve.

Regarding the documentation, the expectation is that if you've read through the entire toolchain introduction and the build system introduction chapter, then you should have a fairly good understanding of the underlying conceptual model and key mechanisms. So perhaps one way to improve the documentation would be for someone (ideally more than one person) to read through both while making note of things that are still unclear/confusing. We could then try to clarify things, add additional notes, etc.

**Comment by boris-kolpackov** [comment]
Also, the more I think about it the more I keep coming back to the analogy of axioms in math: what we are trying to do in `build2` is come up with an underlying model (a set of axioms) that is general and powerful enough to handle most real-world build requirements. So maybe it would be helpful to explicitly and succinctly document the list of `build2` "axioms", perhaps with links to parts of the documentation where relevant things are discussed in more detail? Maybe having a relatively short summary which one can read over several times will help with absorbing the conceptual model?

**Comment by Klaim** [comment]
The documentation I made for my previous colleagues as a "quick intro to build2" starts with a general overview of the conceptual model (without entering into too much details). In particular, explaining what a configuration is is crucial for people to understand the rest of how `build2` works, or even the rest of the documentation. I explain the model by giving some idea of what each important term means.
Then it briefly explains "configuration vs project vs package" - again without detailing these, just stating that these concepts can overlap and enable features. This is necessary to clarify these terms because otherwise it makes the rest of the document very hard to understand.
Then it clarifies what's the role of the files and directories generated by default `build2` projects - this is to prevent the overwhelming feeling the first time people try `bdep new` (from feedback I got, they almost "panic" because it generates a lot of files and directories).
Finally it gives a short overview of the `buildfile` syntax, in particular the core of the grammar, variables and types, and targets, so that people don't get too lost in basics.
On the side of this document I wrote a "short" FAQ for the usual cases (how to build & test, how to add/remove dependencies, how to release, why not use X or Y, etc.) and the usual questions I got from that team when I explained `build2` in the past.

(Both document goes over these things superficially and keep pointing at the official documentation for getting details)

From what my colleagues said, it was far clearer than previous attempts at introducing `build2`, though it's a very small sample of people. Also, I wrote it with them in mind, so I don't know if it's useful in general. It might contain misunderstandings on my part too.

But in my experience it's very hard to explain successfully `build2` without going in this order. So I agree that an introduction to the conceptual model is very important.
Personally I prefer to explain it up-front before explaining anything else, but some people prefer to experience the tool too in parallel, so I adapt to the people I face.

(Making comparisons with how CMake works also helps a lot with people familiar with it, that is most C++ developers these days).

That being said, I also think the error reports could be better. A lot of error messages have been made far better (and with tips!) through the years, thanks to the team for that 👍 it have been very helpful (for me and my colleagues).
There is still some improvements possible IMO though and I don't believe having a better documentation is enough to really fix that. IMO, both need to be improved. (and I do realize that it's a lot of work, I wish I could help more ^^; - also I would like to see some other features in priority, though thinking like this would always push back documentation improvements...)

**Comment by kamrann** [comment]
It's probably true that people tend to jump in too quickly, but realistically anyone trying out a new build tool is going to be attempting to get their project to build with it. There won't be too many people taking the time to learn it as an academic exercise. So while you can try to encourage people to take it slow, the bottom line is if it's too unfriendly for people jumping straight in, then b2 will have a low take-up rate.

> So perhaps one way to improve the documentation would be for someone (ideally more than one person) to read through both while making note of things that are still unclear/confusing. 

I'll try to do so, but at this point I can't separate my understanding of the section from my knowledge gained up to now, so it may be of limited use. Perhaps a super quick questionnaire embedded at the end of these sections to get some feedback from first-time readers?

> Maybe having a relatively short summary which one can read over several times will help with absorbing the conceptual model?

I definitely like that idea. The intro and build manual are very much of the "sit down and read from start to finish" form, which is fine but there's a lot in there - it's not feasible to fully grasp it all in one read through, nor really to use it as a reference.

> That being said, I also think the error reports could be better. [...] There is still some improvements possible IMO though and I don't believe having a better documentation is enough to really fix that. 

I agree with this. I'm all for more documentation, but if the aim is to increase the number of people using b2, then I'd say this is the key thing. People will jump in, and if they keep running into errors that are hard to comprehend, they will quit.

Since this seems to have become a thread on general errors/comprehension, it seems as good a place as any to list some of the specific things I've run into. I'll dump them in a post below, they can be moved elsewhere if preferred.

**Comment by kamrann** [comment]
- The `no rule to update ...` error can mean a missing `depends` in the package manifest, improper `export.build` setup, probably a lot more besides; more info would be useful.
- `unable to import target ...` is another error I've run into a lot that seems to have a number of causes. Also, from what I understand, the `use config.import.libhello configuration variable` suggestion is perhaps misleading since in most cases this is not really what should be required and the issue is something else.
- It seems a library built dynamic with no exported symbols doesn't produce a dll export lib, resulting in errors about that import lib being missing when trying to build something that links against it. Makes sense, but I think many people will probably run into this like I did, where I wasn't even intending to build dynamic. Of course this can be controlled, but again an earlier warning about this case would perhaps make sense?
- `package/repository is now broken` errors seem to come up with fairly high frequency, which is frustrating. Specifically, this would often come up in response to a simple query command, after having previously executed a mutating command that appeared to run successfully. I suspect this may relate to Windows file system issues, as was mentioned somewhere else by @Klaim (are you still running into them?). Is it possible some I/O failures are going undetected? Having an open Explorer window seemed to trigger this, suggesting that commands that needed to create/delete files were encountering a failure due to file locks, but that was not being reported.
- The `bdep`/`bpkg`/`b` hierarchy seems a nice design but can also be complex and confusing at times. In particular, often when running a `bdep` command that results in a problem, there is a suggestion to run `bpkg` (eg `fetch`). It's not clear if this is because it specifically needs to be done at a lower level, or if it's just a case of `bdep` having invoked `bpkg` and therefore the command output/error message is coming from the latter, rather than from the command invoked directly by the user.

Regarding catching the root cause as opposed to just the end result encountered by the build; obviously this is hard and in some cases impossible, since as you say it's a general tool and can't assume too much. But I think even just adding some hint output listing common things that can lead to the issue would be a big step forward. This could obviously be controlled by a user option, but I think it would be good to have more hints on by default for new users.

**Comment by alexchandel** [comment]
> So maybe it would be helpful to explicitly and succinctly document the list of build2 "axioms", perhaps with links to parts of the documentation where relevant things are discussed in more detail? Maybe having a relatively short summary which one can read over several times will help with absorbing the conceptual model?

🙌 Yes 🙌

That, a little more exhaustive documentation of the buildfile language / lexicon, and (stretch goal) smarter error messages with hints would really flatten the learning curve.
